File: api-java-27/javaSdkDemoApp/src/main/java/com/cb/javaSdkDemo/repository/SdkDemoRepository.java
package com.cb.javaSdkDemo.repository;

import com.couchbase.client.java.CouchbaseCluster;
import com.couchbase.client.java.document.JsonDocument;
import com.couchbase.client.java.document.json.JsonObject;
import com.couchbase.client.java.document.json.JsonArray;
import com.couchbase.client.core.message.kv.subdoc.multi.Lookup;
import com.couchbase.client.core.message.kv.subdoc.multi.Mutation;
import com.couchbase.client.java.Bucket;
import com.couchbase.client.java.query.N1qlQuery;
import com.couchbase.client.java.query.N1qlParams;
import com.couchbase.client.java.query.N1qlQueryResult;
import com.couchbase.client.java.query.N1qlQueryRow;
import com.couchbase.client.java.subdoc.DocumentFragment;
/*FTS*/
import com.couchbase.client.java.search.*;
import com.couchbase.client.java.search.queries.*;
import com.couchbase.client.java.search.result.SearchQueryResult;
import com.couchbase.client.java.search.result.SearchQueryRow;

import rx.Observable;
import rx.functions.Func1;

import java.util.*;

import org.springframework.dao.DataRetrievalFailureException;

public final class SdkDemoRepository {

    private static SdkDemoRepository INSTANCE;
    // NOTE: if env variables created after VS Code was opened, restart VS code
    private String host;
    private String bucketName;
    private String username;
    private String password;

    private CouchbaseCluster cluster;
    private Bucket bucket;
    private Boolean connected;

    private SdkDemoRepository() {  
        this.connected = false;
    }

    public static SdkDemoRepository getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new SdkDemoRepository();
        }

        return INSTANCE;
    }

    public Boolean Connect(String host, String bucketName, String user, String password) 
        throws Exception {

        this.host = host;
        this.bucketName = bucketName;
        this.username = user;
        this.password = password;

        if (this.connected) {
            this.Disconnect();
        }

        this.cluster = CouchbaseCluster.create(this.host);
        cluster.authenticate(this.username, this.password);

        this.bucket = cluster.openBucket(this.bucketName);

        this.connected = true;

        return this.connected;
    }

    public Boolean Disconnect() throws Exception {
        this.connected = this.cluster.disconnect();
        if(!this.connected){
            throw new Exception("Unable to disconnect from previous connection.");
        }
        return this.connected;
    }

    public List<String> GetSampleDocIds(){
        String localBucket = this.bucketName;
        if(this.bucketName.contains("-")){
            localBucket = "`" + this.bucketName + "`";
        }

        String query = String.format("SELECT meta().id FROM %s LIMIT 5;", localBucket);

        N1qlQueryResult result = this.bucket.query(N1qlQuery.simple(query));

        if (!result.finalSuccess()) {
            throw new DataRetrievalFailureException("Query error: " + result.errors());
        }

        List<String> ids = new ArrayList<String>();
        for (N1qlQueryRow row : result) {
            ids.add(row.value().getString("id"));
        }

        return ids;
    }

    public String bucketName() {
        return this.bucket.name();
    }

    public List<Map<String, Object>> n1qlQuery(String query, Boolean prepare, JsonArray parameters) {

        N1qlParams params = N1qlParams.build().adhoc(prepare);
        N1qlQueryResult result = null;

        if(!parameters.isEmpty()){
            result = this.bucket.query(N1qlQuery.parameterized(query, parameters, params));
        }
        else{
            result = this.bucket.query(N1qlQuery.simple(query, params));
        }

        if (!result.finalSuccess()) {
            throw new DataRetrievalFailureException("Query error: " + result.errors());
        }

        List<Map<String, Object>> data = new ArrayList<Map<String, Object>>();
        for (N1qlQueryRow row : result) {
            data.add(row.value().toMap());
        }

        return data;
    }

    public Map<String, Object> get(String docId){
        JsonDocument doc = bucket.get(docId);

        return new HashMap<String, Object>(){{
            put(doc.id(), doc.content().toMap());
        }};
    }

    public List<Map<String, Object>> getMulti(List<String> docIds){

        List<JsonDocument> documents = Observable
            .from(docIds)
            .flatMap(new Func1<String, Observable<JsonDocument>>(){
                @Override
                public Observable<JsonDocument> call(String id){
                    return bucket.async().get(id);
                }
            })
            .toList()
            .toBlocking()
            .single();

        List<Map<String, Object>> results = new ArrayList<Map<String, Object>>();
        for (JsonDocument doc : documents) {
            results.add(new HashMap<String, Object>(){{
                put(doc.id(), doc.content().toMap());
            }});
        }

        return results;
    }

    public Map<String, Object> getReplica(String docId){

        Iterator<JsonDocument> docs = bucket.getFromReplica(docId);
        
        if(docs.hasNext()){
            JsonDocument doc = docs.next();

            return new HashMap<String, Object>(){{
                put(doc.id(), doc.content().toMap());
            }};
        }

        return new HashMap<String, Object>(){{}};
    }

    public Boolean touch(String docId, Integer expiry){
        
        Boolean success = this.bucket.touch(docId, expiry);
        
        return success;
    }

    public Map<String, Object> getAndTouch(String docId, Integer expiry){
        
        JsonDocument doc = bucket.getAndTouch(docId, expiry);
        
        return new HashMap<String, Object>(){{
            put(doc.id(), doc.content().toMap());
        }};
    }

    public Map<String, Object> upsert(String docId, String doc){
        JsonObject content = JsonObject.fromJson(doc);
        JsonDocument newDoc = JsonDocument.create(docId, content);

        JsonDocument savedDoc = bucket.upsert(newDoc);

        return new HashMap<String, Object>(){{
            put(savedDoc.id(), savedDoc.content().toMap());
        }};
    }

    public Map<String, Object> insert(String docId, String doc){
        JsonObject content = JsonObject.fromJson(doc);
        JsonDocument newDoc = JsonDocument.create(docId, content);

        JsonDocument savedDoc = bucket.insert(newDoc);

        return new HashMap<String, Object>(){{
            put(savedDoc.id(), savedDoc.content().toMap());
        }};
    }

    public Map<String, Object> replace(String docId, String doc){
        JsonObject content = JsonObject.fromJson(doc);
        JsonDocument newDoc = JsonDocument.create(docId, content);

        JsonDocument savedDoc = bucket.replace(newDoc);

        return new HashMap<String, Object>(){{
            put(savedDoc.id(), savedDoc.content().toMap());
        }};
    }

    public String remove(String docId){

        JsonDocument removedDoc = bucket.remove(docId);

        return removedDoc.id();
    }

    public Map<String, Object> lookupIn(String docId, String path, String resultType){
        DocumentFragment<Lookup> result = bucket.lookupIn(docId).get(path).execute();
        
        if(resultType.equalsIgnoreCase("primitive")){
            Object obj = result.content(0);
            return new HashMap<String, Object>(){{
                put(path, obj);
            }};
        }
        else if(resultType.equalsIgnoreCase("object")){
            JsonObject obj = result.content(0, JsonObject.class);
            return new HashMap<String, Object>(){{
                put(path, obj.toMap());
            }};
        }else{
            //TODO: array
            JsonArray array = result.content(0, JsonArray.class);

            return new HashMap<String, Object>(){{
                put(path, array);
            }};
        }
    }

    public Map<String, Object> mutateIn(String docId, String path, String value, String resultType) {
        DocumentFragment<Mutation> result = bucket.mutateIn(docId).upsert(path, value).execute();

        if (result.exists(path)) {
            return lookupIn(docId, path, resultType);
        }

        return new HashMap<String, Object>() {
        };
    }

    public List<Map<String, Object>> fts(String searchTerm, String indexName, Integer fuzzyLevel) {
        if (indexName.isEmpty()) {
            indexName = "default";
        }

        MatchQuery query = SearchQuery.match(searchTerm).fuzziness(fuzzyLevel);

        SearchQueryResult result = bucket.query(new SearchQuery(indexName, query).limit(10).highlight());

        List<Map<String, Object>> data = new ArrayList<Map<String, Object>>();
        for (SearchQueryRow row : result.hits()) {

            data.add(new HashMap<String, Object>() {
                {
                    put("id", row.id());
                    put("hits", row.fragments());
                }
            });
        }

        return data;
    }
}

File: api-node-26/repository/sdkDemoRepository.js
const couchbase = require("couchbase");
const N1qlQuery = require("couchbase").N1qlQuery;
const searchQuery = require("couchbase").SearchQuery;

class SdkRepository {
  constructor() {
    this.host = "";
    this.bucketName = "";
    this.username = "";
    this.password = "";
    this.sampleIds = [];

    this.cluster = null;
    this.bucket = null;
    this.connected = false;
  }

  connect(host, bucket, username, password, callback) {
    this.host = host;
    this.bucketName = bucket;
    this.username = username;
    this.password = password;

    this.cluster = new couchbase.Cluster(this.host);

    this.cluster.authenticate(this.username, this.password);

    let self = this;
    this.bucket = this.cluster.openBucket(this.bucketName, function(
      err,
      result
    ) {
      if (err) {
        self.connected = false;
      } else {
        self.connected = true;
      }
      callback(err, result);
    });
  }

  getSampleDocIds(callback) {
    let bucket = this.bucketName;
    if (bucket.includes("-")) {
      bucket = "`" + this.bucketName + "`";
    }

    let query = N1qlQuery.fromString(
      "SELECT meta().id FROM " + bucket + " LIMIT 5;"
    );

    let self = this;
    this.bucket.query(query, function(err, rows) {
      if (!err) {
        self.sampleIds = [];
        for (var row in rows) {
          self.sampleIds.push(rows[row].id);
        }
      }
      callback(err, self.sampleIds);
    });
  }

  n1qlQuery(query, prepare, parameters, callback) {
    let n1qlQuery = N1qlQuery.fromString(query);

    if (prepare != null && prepare) {
      n1qlQuery.adhoc(true);
    }

    this.bucket.query(n1qlQuery, parameters, function(err, rows) {
      let results = [];
      if (!err) {
        for (var row in rows) {
          results.push(rows[row]);
        }
      }
      callback(err, results);
    });
  }

  get(docId, callback) {
    this.bucket.get(docId, function(err, doc) {
      let document = null;
      if (!err) {
        document = doc;
      }
      callback(err, document);
    });
  }

  getMulti(docIds, callback) {
    let results = [];
    this.bucket.getMulti(docIds, function(err, docs) {
      if (!err) {
        for (var key in docs) {
          if (docs[key].error) {
            continue;
          }
          results.push(docs[key]);
        }
      }
      callback(err, results);
    });
  }

  getReplica(docId, callback) {
    this.bucket.getReplica(docId, function(err, doc) {
      let document = null;
      if (!err) {
        document = doc;
      }
      callback(err, document);
    });
  }

  touch(docId, expiry, callback) {
    this.bucket.touch(docId, expiry, function(err, result) {
      callback(err, result);
    });
  }

  getAndTouch(docId, expiry, callback) {
    this.bucket.getAndTouch(docId, expiry, function(err, doc) {
      let document = null;
      if (!err) {
        document = doc;
      }
      callback(err, document);
    });
  }

  upsert(docId, docValue, options, callback) {
    let self = this;
    this.bucket.upsert(docId, docValue, options, function(err, result) {
      if (!err) {
        self.get(docId, callback);
      } else {
        callback(err, result);
      }
    });
  }

  insert(docId, docValue, options, callback) {
    let self = this;
    this.bucket.insert(docId, docValue, options, function(err, result) {
      if (!err) {
        self.get(docId, callback);
      } else {
        callback(err, result);
      }
    });
  }

  replace(docId, docValue, options, callback) {
    let self = this;
    this.bucket.replace(docId, docValue, options, function(err, result) {
      if (!err) {
        self.get(docId, callback);
      } else {
        callback(err, result);
      }
    });
  }

  remove(docId, callback) {
    this.bucket.remove(docId, function(err, result) {
      if (!err) {
        callback(err, docId);
      } else {
        callback(err, result);
      }
    });
  }

  lookupIn(docId, path, callback) {
    this.bucket
      .lookupIn(docId)
      .get(path)
      .execute(function(err, result) {
        let subDoc = null;
        if (!err) {
          subDoc = {
            path: path,
            result: result.content(path)
          };
        }
        callback(err, subDoc);
      });
  }

  mutateIn(docId, path, value, callback) {
    this.bucket
      .mutateIn(docId)
      .upsert(path, value)
      .execute(function(err, result) {
        let subDoc = null;
        if (!err) {
          subDoc = {
            path: path,
            success: result.content(path) ? true : false
          };
        }
        callback(err, subDoc);
      });
  }

  fts(term, indexName, fuzziness, callback) {
    if (indexName == null || indexName === "") {
      indexName = "default";
    }

    let match =
      fuzziness != null
        ? searchQuery.term(term).fuzziness(fuzziness)
        : searchQuery.term(term);

    let query = searchQuery
      .new(indexName, match)
      .limit(10)
      .highlight();

    this.bucket.query(query, function(err, res, meta) {
      let results = [];
      if (!err) {
        for (let i = 0; i < res.length; i++) {
          results.push({
            id: res[i].id,
            hit: res[i].locations
          });
        }
      }
      callback(err, results);
    });
  }

  disconnect() {
    this.bucket.disconnect();
    this.bucket = null;
    this.cluster = null;
  }
}

module.exports = new SdkRepository();
